File: 0874-backspace-string-compare/0874-backspace-string-compare.rb
def backspace_compare(s, t)
    i, j = s.length - 1, t.length - 1
    skip_s, skip_t = 0, 0

    while i >= 0 || j >= 0
        while i >= 0
            if s[i] == '#'
                skip_s += 1
                i -= 1
            elsif skip_s > 0
                skip_s -= 1
                i -= 1
            else
                break
            end
        end

        while j >= 0
            if t[j] == '#'
                skip_t += 1
                j -= 1
            elsif skip_t > 0
                skip_t -= 1
                j -= 1
            else
                break
            end
        end

        if i >= 0 && j >= 0 && s[i] != t[j]
            return false
        end

        if (i >= 0) != (j >= 0)
            return false
        end

        i -= 1
        j -= 1
    end

    return true
end

File: 0874-backspace-string-compare/0874-backspace-string-compare.cs
public class Solution {
    public bool BackspaceCompare(string S, string T) {
        int i = S.Length - 1, j = T.Length - 1;
        int skipS = 0, skipT = 0;

        while (i >= 0 || j >= 0) {
            while (i >= 0) {
                if (S[i] == '#') {
                    skipS++;
                    i--;
                } else if (skipS > 0) {
                    skipS--;
                    i--;
                } else {
                    break;
                }
            }

            while (j >= 0) {
                if (T[j] == '#') {
                    skipT++;
                    j--;
                } else if (skipT > 0) {
                    skipT--;
                    j--;
                } else {
                    break;
                }
            }

            if (i >= 0 && j >= 0 && S[i] != T[j]) {
                return false;
            }

            if (i >= 0 != j >= 0) {
                return false;
            }

            i--;
            j--;
        }

        return true;
    }
}